### PR TITLE
Create new TreeOfBoxes on refinement

### PR DESCRIPTION
**Pull request by alexfikl**
Is this what you had in mind in https://github.com/inducer/boxtree/pull/71#issuecomment-1372598338?

### CODE REVIEW

**Comment by alexfikl** [comment]
Hm, can't set `box_flags` and `level_start_box_nrs` to None, since that would fail the traversal tests.

From what I understand in that test is that it refines `nlevels` times and then does the traversal, but before the `box_flags = [1]`. How did that ever work? It seems like the traversal tries to access `box_flags[box_id]`, which should be out of bounds..

**Comment by inducer** [comment]
Yes, I like this a whole lot better. Let me look into making it work.

**Comment by inducer** [comment]
This mostly seems to work. How do you feel about this?

File: boxtree/distributed/local_tree.py [hunk 1]
```diff
@@ -155,24 +155,18 @@ def modify_target_flags_kernel(self):
...
-                IS_TARGET_BOX=(
-                    "(" + box_flag_t + ") " + str(box_flags_enum.IS_TARGET_BOX)
-                ),
-                HAS_TARGET_CHILD_BOXES=(
-                    "(" + box_flag_t + ") " + str(box_flags_enum.HAS_TARGET_CHILD_BOXES)
-                )
-            )
+                    box_flags[i] |= HAS_TARGET_CHILD_BOXES;
+            """,
+            preamble=box_flags_enum.get_c_defines()
```
Comment (inducer) [inline boxtree/distributed/local_tree.py:169]: FYI @gaohao95

**Comment by alexfikl** [comment]
> This mostly seems to work. How do you feel about this?

It looks good to me! Don't think I have any other worries about those attributes getting updated and the rest seems ok.

Thanks for all you help!

**Comment by inducer** [comment]
Thanks for taking a look!

**Pull request merged**